File: 2022/07/java/Main.java
import java.io.BufferedReader;
import java.io.FileReader;
import java.io.IOException;
import java.util.ArrayList;
import java.util.HashMap;
import java.util.List;
import java.util.Map;

public class Main {
	// --- Constants ---
	private static final String INPUT_FILE = "../input/input.txt";
	private static final int LIMIT = 100_000;
	private static final int DISK_SIZE = 70_000_000;
	private static final int UPDATE_SIZE = 30_000_000;

	/**
	 * This private class represents a node in the file system tree.
	 */
	private static abstract class Node {
		protected String name;
		protected int size;
		protected Node parent;
		protected Map<String, Node> childs;

		private Node(Node parent, String name) {
			this.name = name;
			this.size = 0;
			this.parent = parent;
			this.childs = null;
		}
	}

	/**
	 * This private class represents a directory in the file system tree.
	 */
	private static class NodeDir extends Node {
		private NodeDir(NodeDir parent, String name) {
			super(parent, name);
			super.childs = new HashMap<>();
		}
	}

	/**
	 * This private class represents a file in the file system tree.
	 */
	private static class NodeFile extends Node {
		private NodeFile(NodeDir parent, String name, int size) {
			super(parent, name);
			super.size = size;
		}
	}

	/**
	 * This recursive method updates all the directories sizes according to their content.
	 */
	private static void updateDirSizes(NodeDir node) {
		for (Node c: node.childs.values()) {
			if (c instanceof NodeDir) updateDirSizes((NodeDir) c);
			node.size += c.size;
		}
	}

	/**
	 * This recursive method searches all directories smaller in size than a limit.
	 */
	private static int sumSmallDirs(NodeDir node, int limit) {
		int total = 0;
		for (Node c: node.childs.values()) {
			if (c instanceof NodeDir) {
				if (c.size <= limit) total += c.size;
				total += sumSmallDirs((NodeDir) c, limit);
			}
		}
		return total;
	}

	/**
	 * This recursive method searches for the smaller directory over a limit.
	 */
	private static int findDirToDelete(NodeDir node, int need) {
		int result = -1;
		for (Node c: node.childs.values()) {
			if (c instanceof NodeDir) {
				int i = findDirToDelete((NodeDir) c, need);
				if (i != -1 && (result == -1 || result > i)) result = i;
			}
			if (c.size >= need && (result == -1 || result > c.size)) result = c.size;
		}
		return result;
	}

	/**
	 * This recursive method prints a node and every child it has.
	 */
	private static void printNode(Node node, int level) {
		for (int i = 0; i < level; i++) System.out.print("  ");
		System.out.println(node.name + ": " + node.size);
		if (node instanceof NodeDir)
			for (Node c: node.childs.values())
				printNode(c, level + 1);
	}

	public static void main(String[] args) throws IOException {
		NodeDir root = new NodeDir(null, "/");
		NodeDir current = null;

		// --- Reads and parse the input file ---
		try (BufferedReader reader = new BufferedReader(new FileReader(INPUT_FILE))) {
			String line;
			while ((line = reader.readLine()) != null) {
				if (line.startsWith("$ ")) {
					line = line.substring(2);
					if (line.startsWith("cd ")) {
						String name = line.substring(3);
						if (name.equals("/")) current = root;
						else if (name.equals("..")) current = (NodeDir) current.parent;
						else current = (NodeDir) current.childs.get(name);
					} else if (line.equals("ls")) continue;
					else throw new RuntimeException("Unsupported commmand: " + line);
				} else if (line.startsWith("dir ")) {
					String name = line.substring(4);
					Node dir = new NodeDir(current, name);
					current.childs.put(name, dir);
				} else if (line.matches("^\\d+\\s[\\w\\.]+$")) {
					String[] tokens = line.split(" ");
					String name = tokens[1];
					int size = Integer.parseInt(tokens[0]);
					Node file = new NodeFile(current, name, size);
					current.childs.put(name, file);
				}
			}
		}
		updateDirSizes(root);

		// --- Puzzle 1 ---
		int total = sumSmallDirs(root, LIMIT);
		System.out.format("1. The sum of the total sizes of the directories under %,d is %,d%n", LIMIT, total);

		// --- Puzzle 2 ---
		int need = UPDATE_SIZE - (DISK_SIZE - root.size);
		int size = findDirToDelete(root, need);
		System.out.format("2. The total size of the smalles directory needed is %,d%n",  size);
	}
}

File: 2022/07/c/main.c
#include <limits.h>
#include <stdio.h>
#include <stdlib.h>
#include <string.h>
#include <regex.h>
#include <file_reader.h>
#include <string_utils.h>

// --- Constants ---
#define INPUT_FILE "../input/input.txt"
#define MAX_CHILDS 25
#define LIMIT 100000UL
#define DISK_SIZE 70000000UL
#define UPDATE_SIZE 30000000UL

// --- Structs ---
READ_BY_LINE_HEAD(file_t)

struct node {
	struct node* parent;
	char* name;
	unsigned long size;
	unsigned int child_count;
	struct node* childs[MAX_CHILDS];
};

// Creates the root node
struct node* create_root() {
	struct node* n = malloc(sizeof(struct node));
	if (n == NULL) {
		fprintf(stderr, "Malloc failed!");
		exit(EXIT_FAILURE);
	}
	n->parent = NULL;
	n->name = strdup("[root]");
	n->size = 0;
	n->child_count = 0;
	return n;
}

// Adds a child node to another
struct node* add_child(struct node* parent, char* name, unsigned long size) {
	if (parent->child_count == MAX_CHILDS) {
		fprintf(stderr, "Child overflow!");
		exit(EXIT_FAILURE);
	}

	struct node* n = malloc(sizeof(struct node));
	if (n == NULL) {
		fprintf(stderr, "Malloc failed!");
		exit(EXIT_FAILURE);
	}

	n->parent = parent;
	n->name = strdup(name);
	n->size = size;
	n->child_count = 0;

	parent->childs[parent->child_count++] = n;
	return n;
}

// Free the moemory used by a node and all its childrens
void do_free(struct node* n) {
	int i; for (i = 0; i < n->child_count; i++) {
		do_free(n->childs[i]);
		n->childs[i] = NULL;
	}
	if (n->name != NULL) free(n->name);
	free(n);
}

// Private recursive method that prints the tree
void _do_print(struct node* n, int ind) {
	int i; for (i = 0; i < ind; i++) printf("-");
	printf("%s", n->name);
	if (n->size > 0) printf(": %lu", n->size);
	printf("\n");
	for (i = 0; i < n->child_count; i++) _do_print(n->childs[i], ind + 1);
}

// Prints the tree
void print(struct node* n) {
	_do_print(n, 0);
}

// Helper function that selects the smallest number
unsigned long minul(unsigned long a, unsigned long b) {
		if (a < b) return a;
		return b;
}

// Update the tree with the sizes of the directories summing its content
unsigned long update_dir_sizes(struct node* n) {
	if (n->child_count == 0) {
		return n->size;
	}

	unsigned long total = 0;
	int i; for (i = 0; i < n->child_count; i++) total += update_dir_sizes(n->childs[i]);
	n->size = total;
	return total;
}

// Sum used in the puzzle 1
unsigned long sum_puzzle_1(struct node* n) {
	if (n->child_count == 0) {
		return 0;
	}
	unsigned long total = 0;
	int i; for (i = 0; i < n->child_count; i++) total += sum_puzzle_1(n->childs[i]);
	if (n->size <= LIMIT) total += n->size;
	return total;
}

// Search used in the puzzle 2
unsigned long find_puzzle_2(struct node* n, unsigned long need) {
	if (n->size < need || n->child_count == 0) {
		return ULONG_MAX;
	}
	unsigned long min = ULONG_MAX;
	int i; for (i = 0; i < n->child_count; i++) {
		min = minul(min, find_puzzle_2(n->childs[i], need));
	}
	return  minul(min, n->size);
}

// --- Main function ---
int main(void) {
	// --- Regext ---
	regex_t regex;
	int rc = regcomp(&regex, "^[0-9]+.*$", REG_EXTENDED);
	if (rc != 0) {
		fprintf(stderr, "Could not compile regex");
		exit(EXIT_FAILURE);
	}

	// --- Variables ---
	READ_BY_LINE_INIT(file, file_t, INPUT_FILE)
	struct node* root = create_root();
	struct node* ptr = root;
	char line[file.max];
	char buffer[file.max];

	// --- Reads and parse the input file ---
	READ_BY_LINE_WHILE(file, line)
		if (strcmp(line, "$ cd /") == 0) continue;
		if (starts_with("$ cd ", line)) {
			sscanf(line, "$ cd %s", buffer);
			if (strcmp(buffer, "..") == 0) {
				ptr = ptr->parent;
			} else {
				ptr = add_child(ptr, buffer, 0);
			}
		} else if (regexec(&regex, line, 0, NULL, 0) == 0) {
			int size;
			sscanf(line, "%d %s", &size, buffer);
			add_child(ptr, buffer, size);
		}
	READ_BY_LINE_DONE(file)
	regfree(&regex);

	// --- Update the tree dir sizes ---
	update_dir_sizes(root);

	// --- Puzzle 1 ---
	unsigned long total = sum_puzzle_1(root);
	printf("1. The sum of the total sizes of the directories under %lu is %lu\n", LIMIT, total);

	// --- Puzzle 2 ---
	unsigned long need = UPDATE_SIZE - DISK_SIZE + root->size;
	total = find_puzzle_2(root, need);
	printf("2. The total size of the smalles directory needed is %lu\n", total);

	// --- Free resources ---
	do_free(root);
	ptr = NULL;
	root = NULL;

	return EXIT_SUCCESS;
}
